### ISSUE TITLE
Webkit crashes on every other article

### ISSUE THREAD

**Issue by arjan-s** [jangernert/FeedReader#327]
NextCloud: 11.0.1
NextCloud News: 10.1.0

When using the current HEAD version (commit b416761e), webkit crashes on every other article I try to view. The fist article works fine, the second crashes, the third works fine, the fourth crashes, etc.
The crash looks like this:
![image](https://cloud.githubusercontent.com/assets/10400299/22144682/d5589b7a-defe-11e6-8759-e796769ad93a.png)

I'll try to investigate myself, but I'm just dropping it here so you know it happens.

Relevant lines from ui.log:
```
[ DEBUG ] ArticleView: load article 140508
[ DEBUG ] HeaderBar: showArticleButtons true
[ DEBUG ] dbBase.read_article(): 140508
[ DEBUG ] ContentPage: set headerbar
[ DEBUG ] ArticleView: WebView load html
[ DEBUG ] ArticleView: view1 -> view2
[ DEBUG ] ArticleView: loading 10 %
[ DEBUG ] ArticleView: Policy decision
[ INFO  ] ArticleView: webextension-message: on_page_created
[ DEBUG ] ArticleView: load STARTED
[ DEBUG ] DBusConnection: updateFeedList
[ ERROR ] ArticleView: webview crashed
[ DEBUG ] ArticleView: loading 10 %
[ ERROR ] ArticleView: webview crashed
[ DEBUG ] ArticleView: loading 0 %
```

The button "view HTML-code" shows nothing out of the ordinary:
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body><p>Peter Hutterer has announced the latest stable version of libinput, the input handling library commonly used across modern X.Org / Wayland / Mir systems...</p></body></html>
```

**Comment by jangernert** [comment]
Does this happen every time you start feedreader ?
The log only shows crashes, no successfully loaded articles.

**Comment by arjan-s** [comment]
Yes, it always happens, no matter how often I restart the client (and daemon), how many unread articles I have, or how many times I switch between feeds and categories.
I have used git bisect between the current master and v1.6.2, and found that commit ba81e2da6022f51871dba48f8c4c3dad417c0cf3 is the first bad commit. Starting with that commit, I get this bug.
The commit is too invasive to simply revert, and it's too big to immediately see what the problem is, but at least it limits the amount of code to check.
I hope this helps. When I have some more time I'll try to figure out what part of the code actually introduces this problem.

**Comment by jangernert** [comment]
Thx, I already was pretty sure this is where the problem was coming from.
Tonight I'll fix a smaller issue I see right away with the code. Maybe that'll fix your issue. If not we have to investigate further.

**Comment by jangernert** [comment]
Okay, I think I'll at least need the full log to figure this out.

**Comment by arjan-s** [comment]
No problem.
The issue persists whatever my settings are. I have tried changing all of them, one by one, but none of them changed this behaviour.
Even when I just keep randomly changing the theme from the settings, I see the article in the background switching between "WebKit has crashed" and the actual article (with the theme chosen at that moment).
Mouse/keyboard doesn't matter, it crashes with both. Mode doesn't matter. This happens in All, Unread and Starred equally.

These are my full logs:
[daemon.log.txt](https://github.com/jangernert/FeedReader/files/721349/daemon.log.txt)
[ui.log.txt](https://github.com/jangernert/FeedReader/files/721348/ui.log.txt)

To create these logs, I removed the old logs, started FeedReader (in Unread mode by default), clicked one category with unread articles, then started clicking about 10 articles starting from the top. As you can see in the log, the first one succeeded, the second crashed, the third succeeded, and so on. I clicked them with a few seconds pause in between, just enough to read the title and scan the contents.

**Comment by jangernert** [comment]
What's the version of webkitgtk on your system?

**Comment by arjan-s** [comment]
Running on Arch Linux, I have a package called webkit2gtk version 2.14.3-1

**Comment by jangernert** [comment]
Same version I am on as well with no crashes at all.
Anything else @arjan-s and @fbruetting have in common? Wayland vs Xorg? Nvidia drivers? Anything else?

**Comment by JaBecker** [comment]
The same error occurs on my system, also an Arch Linux machine running gnome 3 with intel graphic drivers.

**Comment by arjan-s** [comment]
My system runs Arch, Xorg, Nvidia drivers, XFce

**Comment by jangernert** [comment]
I'm guessing *binary* Nvidia drivers, not the mesa ones
No issues with radeonSI drivers on Fedora 25.
I'm just guessing it could be a driver thing because of the change-log of webkitgtk 2.14.3

> Create GLX OpenGL contexts using version 3.2 (core profile) when available to reduce the memory consumption on Mesa based drivers.

**Comment by fbruetting** [comment]
• I'm running X since Wayland didn't start Gnome some days ago.
• Dual GPU but neither nouveau nor this dual-GPU-Fedora-internal thing (yet) active.
• Dell XPS 15.
• Latest Fedora 25.
• Using some stuff of RPM Fusion and UnitedRPMs, but I don't know how to list them.
• Tomorrow I'll try if Wayland works again and without all shell extensions.

**Comment by jangernert** [comment]
You could try to downgrade WebKit to 2.14.2 or something ;)

**Comment by fbruetting** [comment]
Ok, I'll try.

**Comment by jangernert** [comment]
@fbruetting did you try to downgrade the package yet?

**Comment by arjan-s** [comment]
@jangernert Downgrading webkit2gtk to version 2.14.2 solves the issue for me! I'll block the package upgrade for now, but do you know of a way to investigate/solve this?

**Comment by nbrownuk** [comment]
I'm using Arch Linux and suffered the same problem as @arjan-s. Downgrading from webkit2gtk 2.14.3 to 2.14.2 worked for me, too. It would be great to get a fix for this problem, as 2.14.3 closes a number of critical CVE's. Happy to provide any info about my system, if needed.

**Comment by shemgp** [comment]
Just want to let you know that I'm experiencing the said crashes in Ubuntu Zesty Zapus (development branch) too.  It has 2.14.3-1 of libwebkit2gtk-4.0-37.  Can't downgrade to 2.14.2 as it's not in the repositories.  Tested it by building from git and using checkinstall.  Using flatpak for now until this is fixed. Thank you.

**Comment by jangernert** [comment]
So can we conclude that all of you are on intel-graphics? Anyone on Nvidia or AMD with this issue?

There is a number of things you can do to narrow this down:
1. Log into the Wayland session of gnome and see if this fixes it (I suspect the issue to be glx related)
2. Attach gdb to the web-process and try to get a backtrace of the crash. This could be useful to later on open a bug-report in the WebKit tracker. http://trac.webkit.org/wiki/WebKitGTK/Debugging
3. If you have some cpu-cycles and ram t spare you could track down the commit the issue first occured. I'M suspecting g r210155 or r210154. http://trac.webkit.org/changeset/210154 http://trac.webkit.org/changeset/210155

**Comment by arjan-s** [comment]
Binary Nvidia driver here. I'll try to debug it further anyway.

**Comment by nbrownuk** [comment]
VGA compatible controller: Intel Corporation Iris Graphics 540

**Comment by fbruetting** [comment]
Works on wayland, but with the result below on my 4K monitor. I experienced some likewise behaviour in Chrome apps, that looked like 4k is passed to the applications now (they displayed everything too small, but a program restart resolved this).

You further see a new menu bar if the top bar menu is hidden, but this is already accessible by the program icon top left.

![bildschirmfoto von 2017-01-25 19-18-28](https://cloud.githubusercontent.com/assets/10520767/22303361/43a9573a-e333-11e6-88b9-745791746a8d.png)

**Comment by jangernert** [comment]
Then that's most likely some issue with webkit itself on wayland. Please report this to the webkit-tracker and stay on topic in this report.
So bottom line is: it works in wayland! That makes me think even more it has to be related to the glx-bugfix in webkitgtk 2.14.3
Is anyone willing to build webkit from source and test said commit and one revision before it? Or provide a backtrace of the crash?

**Comment by fbruetting** [comment]
Now WebKitGtk have their own Bugzilla… I surely don't register for every fucking project, just because they coded shit and live in their own goddamn universe, sorry.

**Comment by jangernert** [comment]
Then don't complain to me either :P

**Comment by fbruetting** [comment]
Didn't know whom to blame. ;)

**Comment by jangernert** [comment]
Well, since it is obviously too much to ask to take 5 minutes to register to the webkit-bugtracker, I myself opened a new bug there:
https://bugs.webkit.org/show_bug.cgi?id=167470
However someone of you needs to to a little work and at least get a backtrace, so the devs can start working on the issue:

> I looked through the GitHub issue but I don't see a backtrace there. You need to post a backtrace of the crash with debug symbols for relevant stack frames for us to investigate the issue.

**Comment by arjan-s** [comment]
@jangernert I will try to test it this weekend. Just haven't had the time yet.

**Comment by jangernert** [comment]
@arjan-s great :)
btw:

Carlos Alberto Lopez Perez:
> One workaround is to disable accelerated compositing mode.
You can do that at runtime by exporting the environment variable WEBKIT_DISABLE_COMPOSITING_MODE=1 before starting FeedReader.

**Comment by jangernert** [comment]
> So without a backtrace there's obviously not anything we can do, hence I'm going to close this, sorry. But if you can get a backtrace for this, please do comment or reopen and we'll definitely take a look. There are not so many commits between 2.12.2 and 2.12.3 so it should not be too hard to figure out what went wrong once we have a backtrace.

**Comment by bil-elmoussaoui** [comment]
I can reproduce this bug too. Trying with `WEBKIT_DISABLE_COMPOSITING_MODE=1`  fixes the issue for me.  I tried to use gdb but nothing works correctly :(
I get this whenever i use gdb
```bash
Temporary breakpoint 1 at 0x4254b0
Starting program: /usr/bin/feedreader 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[New Thread 0x7fffe0d26700 (LWP 19096)]
```

**Comment by jangernert** [comment]
Like I posted earlier:

> Debugging WebKit2
If you want to debug the web process, the simplest way is to connect to it using gdb:
$ gdb -p \<pid of WebKitWebProcess>

https://trac.webkit.org/wiki/WebKitGTK/Debugging

**Comment by clopez** [comment]
> Like I posted earlier:
> 
>     Debugging WebKit2
>     If you want to debug the web process, the simplest way is to connect to it using gdb:
>     $ gdb -p <pid of WebKitWebProcess>
> 
> https://trac.webkit.org/wiki/WebKitGTK/Debugging

I have just documented a way of getting a backtrace from a coredump:
https://trac.webkit.org/wiki/WebKitGTK/Debugging#Gettingabacktracefromacoredump

Maybe you find this easier to do.

**Comment by shemgp** [comment]
Here's a crash report from Ubuntu.
[webkit2gtk-4.0.xz.zip](https://github.com/jangernert/FeedReader/files/736954/webkit2gtk-4.0.xz.zip)

**Comment by arjan-s** [comment]
This is my stack trace. It's without debugging symbols since I can't get the source to build yet. I wanted to put it up here anyway to let you know I'm still working on it.

```
#0  0x00000000407ecec1 in  ()
#1  0x00007f5dc9a09464 in  () at /usr/lib/libnvidia-glcore.so.375.26
#2  0x00007f5dc95e23a8 in  () at /usr/lib/libnvidia-glcore.so.375.26
#3  0x00007f5e2f1bb5a7 in  () at /usr/lib/libwebkit2gtk-4.0.so.37
#4  0x00007f5e2f173fff in  () at /usr/lib/libwebkit2gtk-4.0.so.37
#5  0x00007f5e2e202158 in  () at /usr/lib/libwebkit2gtk-4.0.so.37
#6  0x00007f5e2e20add7 in  () at /usr/lib/libwebkit2gtk-4.0.so.37
#7  0x00007f5e2e2086f5 in  () at /usr/lib/libwebkit2gtk-4.0.so.37
#8  0x00007f5e2b181e05 in WTF::RunLoop::performWork() () at /usr/lib/libjavascriptcoregtk-4.0.so.18
#9  0x00007f5e2b1b8129 in  () at /usr/lib/libjavascriptcoregtk-4.0.so.18
#10 0x00007f5e2b5f643a in g_main_context_dispatch () at /usr/lib/libglib-2.0.so.0
#11 0x00007f5e2b5f67f0 in  () at /usr/lib/libglib-2.0.so.0
#12 0x00007f5e2b5f6b12 in g_main_loop_run () at /usr/lib/libglib-2.0.so.0
#13 0x00007f5e2b1b8a30 in WTF::RunLoop::run() () at /usr/lib/libjavascriptcoregtk-4.0.so.18
#14 0x00007f5e2b1b766e in  () at /usr/lib/libjavascriptcoregtk-4.0.so.18
#15 0x00007f5e2b182e15 in  () at /usr/lib/libjavascriptcoregtk-4.0.so.18
#16 0x00007f5e2b1b5b8a in  () at /usr/lib/libjavascriptcoregtk-4.0.so.18
#17 0x00007f5e29c05454 in start_thread () at /usr/lib/libpthread.so.0
#18 0x00007f5e2d7627df in clone () at /usr/lib/libc.so.6
```

**Comment by arjan-s** [comment]
Nope, 2.15.3 doesn't fix it. I'm going to try to build a version with debug symbols tomorrow.

**Comment by arjan-s** [comment]
Traceback of version 2.14.3 with debugging symbols:

```
#0  0x000000004117be01 in  ()
#1  0x00007fab7740e464 in  () at /usr/lib/libnvidia-glcore.so.375.26
#2  0x00007fab76fe73a8 in  () at /usr/lib/libnvidia-glcore.so.375.26
#3  0x00007fabf53f5787 in WebCore::GraphicsContext3D::drawArrays(unsigned int, int, int) (this=0x7fabe070f900, mode=mode@entry=6, first=first@entry=0, count=count@entry=4)
    at /home/user/aur/webkit2gtk-unstable/src/webkitgtk-2.14.3/Source/WebCore/platform/graphics/opengl/GraphicsContext3DOpenGLCommon.cpp:713
#4  0x00007fabf53ae2eb in WebCore::TextureMapperGL::beginClip(WebCore::TransformationMatrix const&, WebCore::FloatRect const&) (this=0x7fabe07cb220, modelViewMatrix=..., targetRect=...)
    at /home/user/aur/webkit2gtk-unstable/src/webkitgtk-2.14.3/Source/WebCore/platform/graphics/texmap/TextureMapperGL.cpp:700
#5  0x00007fabf44533d2 in WebKit::CoordinatedGraphicsScene::paintToCurrentGLContext(WebCore::TransformationMatrix const&, float, WebCore::FloatRect const&, WebCore::Color const&, bool, WebCore::FloatPoint const&, unsigned int) (this=0x7fabe07843f0, matrix=..., opacity=opacity@entry=1, clipRect=..., backgroundColor=..., drawsBackground=<optimized out>, contentPosition=..., PaintFlags=0)
    at /home/user/aur/webkit2gtk-unstable/src/webkitgtk-2.14.3/Source/WebKit2/Shared/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:101
#6  0x00007fabf445c03b in WebKit::ThreadedCompositor::renderLayerTree() (this=0x7fabe076d9c0)
    at /home/user/aur/webkit2gtk-unstable/src/webkitgtk-2.14.3/Source/WebKit2/Shared/CoordinatedGraphics/threadedcompositor/ThreadedCompositor.cpp:201
#7  0x00007fabf4459945 in WTF::Function<void ()>::operator()() const (this=0x7fabe078beb0) at /home/user/aur/webkit2gtk-unstable/src/webkitgtk-2.14.3/Source/WTF/wtf/Function.h:50
#8  0x00007fabf4459945 in WebKit::CompositingRunLoop::<lambda()>::operator() (__closure=0x7fabe078bea8)
    at /home/user/aur/webkit2gtk-unstable/src/webkitgtk-2.14.3/Source/WebKit2/Shared/CoordinatedGraphics/threadedcompositor/CompositingRunLoop.cpp:136
#9  0x00007fabf4459945 in WTF::Function<void()>::CallableWrapper<WebKit::CompositingRunLoop::performTaskSync(WTF::Function<void()>&&)::<lambda()> >::call(void) (this=0x7fabe078bea0)
    at /home/user/aur/webkit2gtk-unstable/src/webkitgtk-2.14.3/Source/WTF/wtf/Function.h:101
#10 0x00007fabf13d72b5 in WTF::Function<void ()>::operator()() const (this=<synthetic pointer>) at /home/user/aur/webkit2gtk-unstable/src/webkitgtk-2.14.3/Source/WTF/wtf/Function.h:50
#11 0x00007fabf13d72b5 in WTF::RunLoop::performWork() (this=0x7fabe0785000) at /home/user/aur/webkit2gtk-unstable/src/webkitgtk-2.14.3/Source/WTF/wtf/RunLoop.cpp:105
#12 0x00007fabf140d599 in WTF::RunLoop::<lambda(gpointer)>::operator() (__closure=0x0, userData=<optimized out>)
    at /home/user/aur/webkit2gtk-unstable/src/webkitgtk-2.14.3/Source/WTF/wtf/glib/RunLoopGLib.cpp:66
#13 0x00007fabf140d599 in WTF::RunLoop::<lambda(gpointer)>::_FUN(gpointer) () at /home/user/aur/webkit2gtk-unstable/src/webkitgtk-2.14.3/Source/WTF/wtf/glib/RunLoopGLib.cpp:68
#14 0x00007fabf184b43a in g_main_context_dispatch () at /usr/lib/libglib-2.0.so.0
#15 0x00007fabf184b7f0 in  () at /usr/lib/libglib-2.0.so.0
#16 0x00007fabf184bb12 in g_main_loop_run () at /usr/lib/libglib-2.0.so.0
#17 0x00007fabf140dea0 in WTF::RunLoop::run() () at /home/user/aur/webkit2gtk-unstable/src/webkitgtk-2.14.3/Source/WTF/wtf/glib/RunLoopGLib.cpp:94
#18 0x00007fabf140cade in WTF::WorkQueue::<lambda()>::operator() (__closure=0x7fab8d7fec40) at /home/user/aur/webkit2gtk-unstable/src/webkitgtk-2.14.3/Source/WTF/wtf/generic/WorkQueueGeneric.cpp:58
#19 0x00007fabf140cade in std::_Function_handler<void(), WTF::WorkQueue::platformInitialize(char const*, WTF::WorkQueue::Type, WTF::WorkQueue::QOS)::<lambda()> >::_M_invoke(const std::_Any_data &) (__functor=...) at /usr/include/c++/6.3.1/functional:1731
#20 0x00007fabf13d8275 in std::function<void ()>::operator()() const (this=0x7fab8d7fec40) at /usr/include/c++/6.3.1/functional:2127
#21 0x00007fabf13d8275 in WTF::threadEntryPoint(void*) (contextData=0x7fabe07f43c0) at /home/user/aur/webkit2gtk-unstable/src/webkitgtk-2.14.3/Source/WTF/wtf/Threading.cpp:60
#22 0x00007fabf140affa in WTF::wtfThreadEntryPoint(void*) (param=0x7fabe07f95a0) at /home/user/aur/webkit2gtk-unstable/src/webkitgtk-2.14.3/Source/WTF/wtf/ThreadingPthreads.cpp:164
#23 0x00007fabefe60454 in start_thread () at /usr/lib/libpthread.so.0
#24 0x00007fabf39b77df in clone () at /usr/lib/libc.so.6
```

**Comment by jangernert** [comment]
Cool stuff :)
I'll post this to the webkit bug-tracker.

**Comment by arjan-s** [comment]
Thanks!

**Comment by jangernert** [comment]
> The missing stack frames at the top are discouraging, but I know there's not much you can do about that. If someone with a free software graphics driver is able to reproduce this crash, that would make this much more likely to be investigated.

**Comment by jangernert** [comment]
> Thanks for the bt. This was fixed in r211204, see bug #167296. I'll merge that in the stable branch before the next release.

Fix will be out with 2.14.4

**Comment by jangernert** [comment]
2.14.4 was just released: https://webkitgtk.org/2017/02/10/webkitgtk2.14.4-released.html
Can someone confirm the fix as soon as the package lands in their distro (probably arch first)?

**Comment by arjan-s** [comment]
I can confirm that it is fixed on my system after installing 2.14.4-1 on Arch. Thanks!

**Comment by nbrownuk** [comment]
Fixed on mine, also. Thanks for getting to the bottom of the problem, and for a very cool piece of software.
